### DISCUSSION TITLE
WebAssembly support for Chromium?

### DISCUSSION THREAD

**Discussion by MasterKia** [uBlockOrigin/uBlock-issues#2891] [General]
https://github.com/gorhill/uBlock/wiki/uBlock-Origin-works-best-on-Firefox#webassembly

https://github.com/WebAssembly/content-security-policy/issues/7#issuecomment-441259729:
> gorhill: However, I do not plan to use `wasm-unsafe-eval` if/when it becomes available, as this goes against the spirit of not executing remote code.

https://github.com/gorhill/uBlock/commit/eb235404bdb1267dd3abd17a47a33312f51723dc:
> Use `wasm-unsafe-eval` instead of deprecated `wasm-eval`

**Comment by gorhill** [comment]
That code is to _detect_ whether `wasm-unsafe-eval` has been added to the `content_security_policy` in the manifest, and if so, then allow wasm.

uBO's `content_security_policy` does not declare `wasm-unsafe-eval`:
https://github.com/gorhill/uBlock/blob/b73a0bef591bb64db81ae2ed5324a775f90a4448/platform/chromium/manifest.json#L79

It's strictly a convenience to easily enable wasm _locally_ to test wasm in Chromium-based browsers when needed (like benchmarking to find out if wasm helps in Chromium-based browser).

**Reply by gorhill** [reply]
To be clear, whoever wants to try uBO with wasm enabled in Chromium, just clone the repo, `cd` into it, then modify [the line](https://github.com/gorhill/uBlock/blob/b73a0bef591bb64db81ae2ed5324a775f90a4448/platform/chromium/manifest.json#L79) from (Chromium's manifest file is at `./platform/chromium/manifest.json`):

    "content_security_policy": "script-src 'self'; object-src 'self'",

To:

    "content_security_policy": "script-src 'self' 'wasm-unsafe-eval'; object-src 'self'",

Then build for Chromium:

    make chromium

Now you have a local build with wasm enabled at `./dist/build/uBlock0.chromium`. Load as unpacked extension in Chromium. The code at [the commit](https://github.com/gorhill/uBlock/commit/eb235404bd) you pointed out above will detect that `'wasm-unsafe-eval'` is part of the manifest's `content_security_policy` property, and thus will allow the wasm-related code paths to be taken in uBO. If you use the debugger pane in the browser tools for uBO, you will see wasm code:

![image](https://github.com/uBlockOrigin/uBlock-issues/assets/585534/fc3b8520-e0d2-421c-93a9-9b75bbcba3b7)

**Reply by MasterKia** [reply]
@gorhill Thanks, added this as an optional step in uBO's build instruction:
https://github.com/MasterKia/uBlock/commit/7a5dc6773d9bd4c8fd35d34c84f2b0c9bb6ccdde
